<a href="https://colab.research.google.com/github/nrao57/tensorflow_hub/blob/main/BirdsTest.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Import and Process the Image to Classify

In [ ]:
from PIL import Image
import numpy as np

Import and Process the image

In [53]:
# Open Image
img_pil = Image.open("/content/bird_images/external-content.duckduckgo.com.jpeg").resize((224, 224), Image.ANTIALIAS)
print("image mode (can be L, RGB, CMYK): " + str(img_pil.mode))
# Cast image to numpy array
img = np.array(img_pil)
print("image shape: " + str(img.shape))
# scale the image to [0, 1]
print("image max value: " + str(img.max()))
img_scaled = np.absolute(img) / img.max()
print("image max value after scaling: " + str(img_scaled.max()))



image mode (can be L, RGB, CMYK): RGB
image shape: (224, 224, 3)
image max value: 255
image max value after scaling: 1.0


In [48]:
# Img after scaling
print(img_scaled)

[[0.496 0.492 0.488 ... 0.532 0.524 0.532]
 [0.488 0.484 0.48  ... 0.516 0.508 0.508]
 [0.476 0.472 0.468 ... 0.5   0.492 0.492]
 ...
 [0.416 0.368 0.48  ... 0.68  0.68  0.692]
 [0.428 0.376 0.516 ... 0.684 0.692 0.696]
 [0.436 0.38  0.548 ... 0.692 0.696 0.708]]


# Import Tensorflow Hub Model as keras layer

[Import tensorflow_hub model as keras layer](https://www.tensorflow.org/hub/tf2_saved_model)




In [ ]:
# TF2 version
import tensorflow.compat.v2 as tf
import tensorflow_hub as hub

m = hub.KerasLayer('https://tfhub.dev/google/aiy/vision/classifier/birds_V1/1')

INFO:tensorflow:Saver not created because there are no variables in the graph to restore


INFO:tensorflow:Saver not created because there are no variables in the graph to restore


Based on the model's documentation, the input to the layer should be 3-channel RGB color images of size 224 x 224, scaled to [0, 1]. And it should be in an array with other images for batch prediction

# Use imported model to make prediction

In [55]:
img_batch = np.array([img_scaled])
probabilities = m(img_batch)
class_ids = tf.argmax(probabilities, axis=-1)

In [58]:
class_ids

<tf.Tensor: shape=(1,), dtype=int64, numpy=array([704])>

since the class_id with the highest probability is 704, we check the label map for the index 704 to see the species of bird

which is:
**Setophaga magnolia**

# Deploy to Production with Vertex AI



1.   Download and Unzip Model from TensorFlow Hub
2.   Create new GCS Cloud Storage Bucket
3.   Upload model to GCS Bucket
4.   Import model into Vertex AI from cloud storage bucket
5.   Deploy to Endpoint
6.   Send requests to endpoint in [this format](https://cloud.google.com/vertex-ai/docs/predictions/online-predictions-custom-models)


# Resources

[TensorFlow Hub for real world impact | Session](https://www.youtube.com/watch?v=BE5nkhFe3AE)

https://medium.com/ymedialabs-innovation/how-to-use-tensorflow-hub-with-code-examples-9100edec29af

https://cloud.google.com/blog/topics/developers-practitioners/serve-tensorflow-model-google-cloud-vertex-ai